File: ts/src/metrics.ts
/**
 * Metrics collector interface + no-op default.
 *
 * See design doc §11 and Go SDK's metrics.go. Implementations must be safe
 * for concurrent use and return quickly — verifiers call these on the request
 * hot path.
 */

import type { PrincipalKind, Verifier as _Verifier } from './verifier.js'
import type { ErrorKind } from './errors.js'

/** VerifyResult tags used for the ObserveVerifyDuration `result` argument. */
export type VerifyResult = 'hit' | 'miss-ok' | 'miss-err' | 'error'

/**
 * MetricsCollector receives internal SDK telemetry. Users typically plug their
 * own Prometheus / OTLP-backed implementation.
 */
export interface MetricsCollector {
  observeVerifyDuration(
    verifier: PrincipalKind,
    result: VerifyResult | string,
    ms: number,
  ): void
  incCacheHit(verifier: PrincipalKind, hit: boolean): void
  incErrorByKind(verifier: PrincipalKind, kind: ErrorKind): void
}

/** noopMetrics discards every observation. Default when Config.metrics is unset. */
export const noopMetrics: MetricsCollector = {
  observeVerifyDuration(_v, _r, _ms) {},
  incCacheHit(_v, _h) {},
  incErrorByKind(_v, _k) {},
}

File: go/metrics.go
package octoauth

import "time"

// MetricsCollector receives internal SDK telemetry. Implementations MUST be
// safe for concurrent use and MUST return quickly; verifiers call these
// methods on the request hot path.
//
// See design doc §11 for the built-in call sites:
//   - ObserveVerifyDuration is called on every Verify with the observed
//     latency and a `result` tag (one of "hit", "miss-ok", "miss-err",
//     "error").
//   - IncCacheHit is called on every cache lookup with hit=true|false.
//   - IncErrorByKind is called whenever a Verify returns a non-nil error.
type MetricsCollector interface {
	ObserveVerifyDuration(verifier PrincipalKind, result string, duration time.Duration)
	IncCacheHit(verifier PrincipalKind, hit bool)
	IncErrorByKind(verifier PrincipalKind, kind ErrorKind)
}

// NoopMetrics is a [MetricsCollector] that discards every observation. It is
// the default when [Config.Metrics] is nil; users plug their own Prometheus
// or OTLP-backed implementation to opt in.
type NoopMetrics struct{}

// ObserveVerifyDuration implements [MetricsCollector].
func (NoopMetrics) ObserveVerifyDuration(_ PrincipalKind, _ string, _ time.Duration) {}

// IncCacheHit implements [MetricsCollector].
func (NoopMetrics) IncCacheHit(_ PrincipalKind, _ bool) {}

// IncErrorByKind implements [MetricsCollector].
func (NoopMetrics) IncErrorByKind(_ PrincipalKind, _ ErrorKind) {}
